src/test/java/io/strimzi/kafka/metrics/KafkaMetricsCollectorTest.java

```diff
@@ -51,35 +50,87 @@ public void testCollect() {
+        Collector.MetricFamilySamples.Sample serverGroupNameSamples1 = metricFamilySamples1.samples.get(0);
+
+        assertEquals("kafka_server_group_name", metricFamilySamples1.name);
+        assertEquals(1, metricFamilySamples1.samples.size());
+        assertEquals(3.0, serverGroupNameSamples1.value, 0.1);
 
         // Removing the metric
         collector.removeMetric(buildMetric("name", "group", 4.0));
         metrics = collector.collect();
         assertTrue(metrics.isEmpty());
     }
 
+    @Test
+    public void testCollectNonNumericMetric() {
+        Map<String, String> props = new HashMap<>();
+        props.put(PrometheusMetricsReporterConfig.ALLOWLIST_CONFIG, "kafka_server_group_name.*");
+        PrometheusMetricsReporterConfig config = new PrometheusMetricsReporterConfig(props);
+        KafkaMetricsCollector collector = new KafkaMetricsCollector(config);
+        collector.setPrefix("kafka.server");
+
+        List<Collector.MetricFamilySamples> metrics = collector.collect();
+        assertTrue(metrics.isEmpty());
+
+        // Adding a non-numeric metric converted
+        KafkaMetric nonNumericMetric = buildNonNumericMetric("name", "group");
+        collector.addMetric(nonNumericMetric);
+        metrics = collector.collect();
+
+        Map<String, String> expectedLabels = new HashMap<>(labels);
+        expectedLabels.put("kafka_server_group_name", "hello");
+        assertMetric(metrics, "kafka_server_group_name", 1.0, expectedLabels);
+        assertEquals(1, metrics.size());
+
+        Collector.MetricFamilySamples metricFamilySamples = metrics.get(0);
+        Collector.MetricFamilySamples.Sample serverGroupNameSamples = metricFamilySamples.samples.get(0);
+
+        assertEquals("kafka_server_group_name", metricFamilySamples.name);
+        assertEquals(1, metricFamilySamples.samples.size());
+        assertEquals(1.0, serverGroupNameSamples.value, 0);
+        assertTrue(serverGroupNameSamples.labelNames.contains("kafka_server_group_name"));
+    }
+
+    public void assertMetric(List<Collector.MetricFamilySamples> metrics, String expectedName, double expectedValue, Map<String, String> expectedLabels) {
+        boolean metricFound = false;
+        for (Collector.MetricFamilySamples metricFamilySamples : metrics) {
+            if (metricFamilySamples.name.equals(expectedName)) {
+                for (Collector.MetricFamilySamples.Sample sample : metricFamilySamples.samples) {
+                    if (sample.value == expectedValue &&
+                            sample.labelNames.equals(new ArrayList<>(expectedLabels.keySet())) &&
+                            sample.labelValues.equals(new ArrayList<>(expectedLabels.values()))) {
+                        metricFound = true;
+                        break;
+                    }
+                }
+            }
+        }
+        assertTrue(metricFound, "Expected metric not found: " + expectedName);
```

Comment:
I don't really like how you've extracted this method.  You want the test to give good feedback when things go wrong so the developer knows precisely what he/she has broken.  The assertTrue masks the detail in an unhelpful way.


This is the kind of assert method I'd have extracted:

 https://github.com/OwenCorrigan76/metrics-reporter/pull/1

Reply:
Implemented as suggested. Thanks for the suggestion